File: daemon/models/connector_status.py
"""
连接器状态枚举 - 修正版设计
分离基本状态和运行状态，更加清晰实用
"""

from enum import Enum
from dataclasses import dataclass
from typing import Dict, Any, Optional
from datetime import datetime


class ConnectorRunningState(Enum):
    """连接器运行状态"""
    STOPPED = "stopped"    # 已停止
    STARTING = "starting"  # 启动中  
    RUNNING = "running"    # 正在运行
    STOPPING = "stopping" # 停止中
    ERROR = "error"        # 错误状态
    
    @classmethod
    def from_string(cls, state: str) -> 'ConnectorRunningState':
        """从字符串创建状态"""
        try:
            return cls(state.lower())
        except ValueError:
            return cls.STOPPED


@dataclass
class ConnectorStatus:
    """
    连接器完整状态信息
    
    设计理念：
    - enabled: 基本状态，是否启用（决定是否自动启动）
    - running_state: 运行状态，表示实际运行情况
    - installed: 虚拟字段，通过数据库存在性计算
    """
    connector_id: str
    display_name: str
    enabled: bool = True                                    # 是否启用（基本状态）
    running_state: ConnectorRunningState = ConnectorRunningState.STOPPED  # 运行状态
    
    # 进程信息
    process_id: Optional[int] = None
    last_heartbeat: Optional[datetime] = None
    
    # 统计信息
    data_count: int = 0
    last_activity: Optional[str] = None
    
    # 错误信息
    error_message: Optional[str] = None
    error_code: Optional[str] = None
    
    @property
    def is_installed(self) -> bool:
        """虚拟字段：是否已安装（通过数据库存在性判断）"""
        # 这个属性在实际使用时会被数据库服务层设置
        return True
    
    @property
    def is_healthy(self) -> bool:
        """是否健康"""
        return self.enabled and self.running_state == ConnectorRunningState.RUNNING
    
    @property
    def should_be_running(self) -> bool:
        """是否应该运行"""
        return self.enabled
    
    def to_dict(self) -> Dict[str, Any]:
        """转换为字典"""
        return {
            "connector_id": self.connector_id,
            "display_name": self.display_name,
            "enabled": self.enabled,
            "running_state": self.running_state.value,
            "is_installed": self.is_installed,
            "is_healthy": self.is_healthy,
            "should_be_running": self.should_be_running,
            "process_id": self.process_id,
            "last_heartbeat": self.last_heartbeat.isoformat() if self.last_heartbeat else None,
            "data_count": self.data_count,
            "last_activity": self.last_activity,
            "error_message": self.error_message,
            "error_code": self.error_code
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'ConnectorStatus':
        """从字典创建状态"""
        return cls(
            connector_id=data["connector_id"],
            display_name=data["display_name"],
            enabled=data.get("enabled", True),
            running_state=ConnectorRunningState.from_string(data.get("running_state", "stopped")),
            process_id=data.get("process_id"),
            last_heartbeat=datetime.fromisoformat(data["last_heartbeat"]) if data.get("last_heartbeat") else None,
            data_count=data.get("data_count", 0),
            last_activity=data.get("last_activity"),
            error_message=data.get("error_message"),
            error_code=data.get("error_code")
        )
    
    def set_error(self, error_message: str, error_code: Optional[str] = None):
        """设置错误状态"""
        self.running_state = ConnectorRunningState.ERROR
        self.error_message = error_message
        self.error_code = error_code
    
    def clear_error(self):
        """清除错误状态"""
        self.error_message = None
        self.error_code = None
        if self.running_state == ConnectorRunningState.ERROR:
            self.running_state = ConnectorRunningState.STOPPED
    
    def update_heartbeat(self):
        """更新心跳时间"""
        self.last_heartbeat = datetime.now()
        if self.running_state == ConnectorRunningState.STARTING:
            self.running_state = ConnectorRunningState.RUNNING


# 工厂函数

def create_new_connector_status(connector_id: str, display_name: str) -> ConnectorStatus:
    """创建新连接器的默认状态"""
    return ConnectorStatus(
        connector_id=connector_id,
        display_name=display_name,
        enabled=True,  # 新连接器默认启用
        running_state=ConnectorRunningState.STOPPED
    )

def create_disabled_connector_status(connector_id: str, display_name: str) -> ConnectorStatus:
    """创建禁用连接器状态"""
    return ConnectorStatus(
        connector_id=connector_id,
        display_name=display_name,
        enabled=False,
        running_state=ConnectorRunningState.STOPPED
    )

File: connectors/shared/src/connector_status.cpp
#include "linch_connector/connector_status.hpp"
#include "linch_connector/unified_client.hpp"
#include <iostream>
#include <unistd.h>

namespace linch_connector {

ConnectorStatusManager::ConnectorStatusManager(const std::string& connector_id, const std::string& display_name) {
    status_.connector_id = connector_id;
    status_.display_name = display_name;
    status_.enabled = true;
    status_.running_state = ConnectorRunningState::STOPPED;
    status_.process_id = getpid();
    last_heartbeat_sent_ = std::chrono::system_clock::now();
}

void ConnectorStatusManager::setState(ConnectorRunningState state) {
    status_.running_state = state;
    status_.clearError(); // 状态改变时清除错误
}

void ConnectorStatusManager::setError(const std::string& error_message, const std::string& error_code) {
    status_.setError(error_message, error_code);
}

void ConnectorStatusManager::setDataCount(int count) {
    status_.data_count = count;
}

void ConnectorStatusManager::setLastActivity(const std::string& activity) {
    status_.last_activity = activity;
}

void ConnectorStatusManager::clearError() {
    status_.clearError();
}

bool ConnectorStatusManager::sendHeartbeat(UnifiedClient& client) {
    auto now = std::chrono::system_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::seconds>(now - last_heartbeat_sent_);
    
    if (duration.count() < HEARTBEAT_INTERVAL_SECONDS) {
        return true; // 还未到发送时间
    }
    
    // 准备心跳数据 - 使用纯IPC格式
    nlohmann::json heartbeat_data = {
        {"connector_id", status_.connector_id},
        {"process_id", status_.process_id},
        {"running_state", RunningStateHelper::toString(status_.running_state)},
        {"data_count", status_.data_count},
        {"timestamp", std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()).count()}
    };
    
    if (!status_.error_message.empty()) {
        heartbeat_data["error_message"] = status_.error_message;
        heartbeat_data["error_code"] = status_.error_code;
    }
    
    bool success = sendIPCUpdate(client, "/heartbeat", heartbeat_data);
    
    if (success) {
        last_heartbeat_sent_ = now;
        status_.updateHeartbeat();
        std::cout << "[INFO] Heartbeat sent successfully for " << status_.connector_id << std::endl;
    } else {
        std::cerr << "[ERROR] Failed to send heartbeat for " << status_.connector_id << std::endl;
    }
    
    return success;
}

bool ConnectorStatusManager::sendStatusUpdate(UnifiedClient& client) {
    // 发送完整状态更新
    nlohmann::json status_data = status_.toIPCJson();
    
    std::string endpoint = "/connectors/" + status_.connector_id + "/status";
    bool success = sendIPCUpdate(client, endpoint, status_data);
    
    if (success) {
        std::cout << "[INFO] Status update sent successfully for " << status_.connector_id << std::endl;
    } else {
        std::cerr << "[ERROR] Failed to send status update for " << status_.connector_id << std::endl;
    }
    
    return success;
}

bool ConnectorStatusManager::notifyStarting(UnifiedClient& client) {
    setState(ConnectorRunningState::STARTING);
    
    nlohmann::json starting_data = {
        {"connector_id", status_.connector_id},
        {"running_state", "starting"},
        {"process_id", status_.process_id},
        {"message", "Connector is starting"}
    };
    
    std::string endpoint = "/connectors/" + status_.connector_id + "/status";
    bool success = sendIPCUpdate(client, endpoint, starting_data);
    
    if (success) {
        std::cout << "[INFO] Starting notification sent for " << status_.connector_id << std::endl;
    }
    
    return success;
}

bool ConnectorStatusManager::notifyStopping(UnifiedClient& client) {
    setState(ConnectorRunningState::STOPPING);
    
    nlohmann::json stopping_data = {
        {"connector_id", status_.connector_id},
        {"running_state", "stopping"},
        {"process_id", status_.process_id},
        {"message", "Connector is stopping"}
    };
    
    std::string endpoint = "/connectors/" + status_.connector_id + "/status";
    bool success = sendIPCUpdate(client, endpoint, stopping_data);
    
    if (success) {
        std::cout << "[INFO] Stopping notification sent for " << status_.connector_id << std::endl;
    }
    
    return success;
}

bool ConnectorStatusManager::sendIPCUpdate(UnifiedClient& client, const std::string& endpoint, const nlohmann::json& data) {
    try {
        // 使用纯IPC协议格式发送
        std::string json_data = data.dump();
        auto response = client.post(endpoint, json_data);
        
        if (response.isSuccess()) { // IPC通信成功
            return true;
        } else {
            std::cerr << "[ERROR] IPC communication failed - " 
                      << response.error_code << ": " << response.error_message << std::endl;
            return false;
        }
    } catch (const std::exception& e) {
        std::cerr << "[ERROR] Exception in sendIPCUpdate: " << e.what() << std::endl;
        return false;
    }
}

// 工厂函数实现
ConnectorStatusV2 createNewConnectorStatus(const std::string& connector_id, const std::string& display_name) {
    ConnectorStatusV2 status;
    status.connector_id = connector_id;
    status.display_name = display_name;
    status.enabled = true;
    status.running_state = ConnectorRunningState::STOPPED;
    status.process_id = getpid();
    return status;
}

} // namespace linch_connector